runtime/parachains/src/dmp.rs

```diff
@@ -149,27 +237,73 @@ impl<T: Config> Pallet<T> {
 		config: &HostConfiguration<T::BlockNumber>,
 		para: ParaId,
 		msg: DownwardMessage,
-	) -> Result<(), QueueDownwardMessageError> {
+	) -> Result<Weight, QueueDownwardMessageError> {
+		// Check if message is oversized.
 		let serialized_len = msg.len() as u32;
 		if serialized_len > config.max_downward_message_size {
 			return Err(QueueDownwardMessageError::ExceedsMaxMessageSize)
 		}
 
+		let mut weight = 0u64;
+		let QueueState { ring_buffer_state, message_window_state } = Self::dmp_queue_state(para);
+		weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 0));
+
+		let mut ring_buf = RingBuffer::with_state(ring_buffer_state, para);
+		let mut message_window = MessageWindow::with_state(message_window_state, para);
+
+		// Check if we need a new page.
+		let mut page_idx = if ring_buf.size() == 0 {
+			// Get a fresh page.
+			ring_buf.extend()
+		} else {
+			// We've checked the queue is not empty, so `last_used` is guaranteed `Some`.
+			// This page might be full but we check that later.
+			ring_buf.last_used().unwrap()
+		};
+
+		let last_used_len = <Self as Store>::DownwardMessageQueuePages::decode_len(page_idx)
+			.unwrap_or(0)
```

Comment:
consider using [`defensive_unwrap_or`](https://github.com/paritytech/substrate/blob/669ed362f922c679ab1416fc8b9e9f4f68a21483/frame/support/src/traits/misc.rs#L97)

Reply:
Coolio, didn't know we have this.